docs/fml-spec.md

```diff
@@ -0,0 +1,576 @@
+
+```yaml
+features:
+  spotlight-search:
+    description: "Configuring how we integrate web-pages with iOS's Spotlight search"
+    variables:
+      # … other variables omitted for clarity
+      item-thumbnail:
+        description: "The icon that appears in the Spotlight search results.
+          Note that changing this does not change already indexed content."
+        type: ThumbnailType
+        default: letter
+types:
+  ThumbnailType:
+    description: An enum of types of icon that can be presented alongside titles of pages
+    variants:
+      letter: A rendering of the first letter of the domain name
+      screenshot: A screenshot thumbnail of the webppage
+      favicon: The favicon derived from the webpage
+      none: No icon is displayed
+```
+
+The value `itemThumbnail` is of type `ThumbnailType`, which is generated as an enum. This can be used exhaustively matched in a `switch` statement.
+
+```swift
+switch spotlightConfig.itemThumbnail {
+    case .favicon:
+        item.thumbnailData = FaviconFetcher.getFaviconFromDiskCache(imageKey: baseDomain)?.pngData()
+    case .letter:
+        item.thumbnailData = FaviconFetcher.letter(forUrl: url).pngData()
+    case .screenshot:
+        item.thumbnailData = tab.screenshot?.pngData()
+    case .none:
+        item.thumbnailData = nil
+}
+```
+
+Note that enumeration variants in Swift are in `mixedCamelCase`; in Kotlin they are in `SCREAMING_SNAKE_CASE`.
+
+## Feature defaults
+
+Already, in this simple example we have three variables. We may want to vary the configuration of the feature when the user is not involved in an experiment.
+
+This might be because we have run an experiment and learned that a change of configuration should be made more permanent.
+
+The `defaults` list is a list of zero or more "default blocks", used to patch the default values of the variables in a feature.
+
+The `value` property defines the patch which is overlaid on top of the existing variable values.
+
+```yaml
+features:
+  spotlight-search:
+    variables:
+      enabled:
+        description: If `false`, the app will not record anything with Spotlight.
+        type: Bool
+        default: false
+      max-age-in-days:
+        description: The number of days a single piece of content is indexed for.
+        type: Double
+        default: 28.0
+      item-thumbnail:
+        description: "The icon that appears in the Spotlight search results.
+          Note that changing this does not change already indexed content."
+        type: ThumbnailType
+        default: letter
+    defaults:
+      - value: { item-thumbnail: screenshot, max-age-in-days: 56.0 }
+```
+
+The `defaults` list of `value`s set the default value of `item-thumbnail` to `screenshot`, and `max-age-in-days` to `56.0`.
+
+Adding new feature blocks to the `defaults` list patches the existing defaults further:
+
+```yaml
+    defaults:
+      - value: { item-thumbnail: screenshot, max-age-in-days: 56.0 }
```

Comment:
q: What is the advantage of keeping the previous default patch? Is it to understand all previously made change when reviewing this? Would it make sense to add a timestamp or versioning to these patches?

Reply:
I think there's a documentary reason (as you've alluded to), but also a speed/accuracy reason: if we wish to make a specific change permanent, then we can copy it exactly in to the manifest, and then let FML sort it out. 

Additionally:

We have talked about tooling to compact/rationalize these patches (i.e. read the manifest, the re-write it after merging these patches).

We have also talked about tooling to download the experiment recipes, then write the rollouts right in to the manifest.